src/org/exist/storage/UpdateBinaryLoggable.java

```diff
@@ -21,89 +34,78 @@
-       byte [] data = new byte[size];
-       in.get(data);
-       original = new File(new String(data));
-       size = in.getInt();
-       data = new byte[size];
-       in.get(data);
-       backup = new File(new String(data));
+    public UpdateBinaryLoggable(final DBBroker broker, final Txn txn, File original, final File backup) {
+        super(NativeBroker.LOG_UPDATE_BINARY, txn.getId());
+        this.original = original;
+        this.backup = backup;
     }
 
-    /* (non-Javadoc)
-     * @see org.exist.storage.log.Loggable#getLogSize()
-     */
+    public UpdateBinaryLoggable(final DBBroker broker, final long transactionId) {
+        super(NativeBroker.LOG_UPDATE_BINARY, transactionId);
+    }
+
+    @Override
+    public void write(final ByteBuffer out) {
+        final String originalPath = original.getAbsolutePath();
+        byte[] data = originalPath.getBytes(StandardCharsets.UTF_8);
+        out.putInt(data.length);
+        out.put(data);
+        final String backupPath = backup.getAbsolutePath();
+        data = backupPath.getBytes(StandardCharsets.UTF_8);
+        out.putInt(data.length);
+        out.put(data);
+    }
+
+    @Override
+    public void read(final ByteBuffer in) {
+        int size = in.getInt();
+        byte[] data = new byte[size];
+        in.get(data);
+        original = new File(new String(data, StandardCharsets.UTF_8));
+        size = in.getInt();
+        data = new byte[size];
+        in.get(data);
+        backup = new File(new String(data, StandardCharsets.UTF_8));
+    }
+
+    @Override
     public int getLogSize() {
-        return 8 + original.getAbsolutePath().getBytes().length + backup.getAbsolutePath().getBytes().length;
+        return 8 + original.getAbsolutePath().getBytes(StandardCharsets.UTF_8).length + backup.getAbsolutePath().getBytes(StandardCharsets.UTF_8).length;
     }
 
+    @Override
     public void redo() throws LogException {
-       // TODO: is there something to do?  The file has been written
+        // TODO: is there something to do?  The file has been written
     }
-    
+
+    @Override
     public void undo() throws LogException {
-       try {
-          final FileInputStream is = new FileInputStream(backup);
-          final FileOutputStream os = new FileOutputStream(original);
-          final byte [] buffer = new byte[4096];
-          int len;
-          while ((len=is.read(buffer))>=0) {
-             os.write(buffer,0,len);
-          }
-          os.close();
-          is.close();
-       } catch (final IOException ex) {
-          
-       }
-    }
-    
-     public String dump() {
-        return super.dump() + " - update "+original+" to "+backup;
+        try(final FileInputStream is = new FileInputStream(backup);
+                final FileOutputStream os = new FileOutputStream(original)) {
+            final byte[] buffer = new byte[4096];
+            int len = -1;
+            while ((len = is.read(buffer)) > -1) {
```

Comment:
copy of bytes might be simplified with library (if new dependancy is good enough for design)